bench/bin/throughput.rs

```diff
@@ -92,6 +99,190 @@ fn start_ingester(
+    stream::unfold(Vec::new(), move |state| async move {
+        let mut state = loop {
+            if !state.is_empty() {
+                break state;
+            }
+
+            // Hitting the metrics endpoint too quick may result in subsequent queries without
+            // any data points. Depending on need, one can specify a delay before scrapping giving
+            // the agent time to generate some new metric data.
+            if let Some(delay) = scrape_delay {
+                tokio::time::sleep(delay).await;
+            }
+
+            match fetch_agent_metrics(metrics_port).await {
+                Ok((StatusCode::OK, Some(body))) => {
+                    let body = body.lines().map(|l| Ok(l.to_string()));
+                    break Scrape::parse(body)
+                        .expect("failed to parse the metrics response")
+                        .samples;
+                }
+                Ok((status, _)) => panic!(
+                    "The /metrics endpoint returned status code {:?}, expected 200.",
+                    status
+                ),
+                Err(err) => panic!(
+                    "Failed to make HTTP request to metrics endpoint - reason: {:?}",
+                    err
+                ),
+            }
+        };
+
+        state.pop().map(|metric| (metric, state))
+    })
+}
+
+#[allow(dead_code)]
+pub struct MetricsRecorder {
+    server: tokio::task::JoinHandle<Vec<Sample>>,
+    abort_handle: AbortHandle,
+}
+
+impl MetricsRecorder {
+    #[allow(dead_code)]
+    pub fn start(port: u16, scrape_delay: Option<Duration>) -> MetricsRecorder {
+        let (abort_handle, abort_registration) = AbortHandle::new_pair();
+        MetricsRecorder {
+            server: tokio::spawn({
+                async move {
+                    Abortable::new(stream_agent_metrics(port, scrape_delay), abort_registration)
+                        .collect::<Vec<Sample>>()
+                        .await
+                }
+            }),
+            abort_handle,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn stop(self) -> Vec<Sample> {
+        self.abort_handle.abort();
+        match self.server.await {
+            Ok(data) => data,
+            Err(e) => panic!("error waiting for thread: {}", e),
+        }
+    }
+}
+
+fn data_pair_for(name: &str) -> impl Fn(&Sample) -> Option<(i64, f64)> + '_ {
+    move |s: &Sample| match (&s.value, &s.labels.get("outcome")) {
+        (Value::Untyped(raw), Some("success")) if s.metric.as_str() == name => {
+            Some((s.timestamp.timestamp_millis(), *raw))
+        }
+        _ => None,
+    }
+}
+// **** IMPORTED **************************
+
+// Sample number of lines
+fn calculate_fs_line_metrics(samples: &[Sample]) -> (f64, f64) {
+    let fs_sample_data = samples
+        .iter()
+        .filter_map(|s| match s.value {
+            Value::Counter(raw) if s.metric.as_str() == "logdna_agent_fs_lines" => Some(raw),
```

Comment:
A good helper function here might encapsulate this string manipulation and property reference so you can write something like:
```
Value::Counter(raw) if is_agent_metric(s, "fs_lines") => Some(raw)
```
since we know that we're always going to prefix "logdna_agent", we always need to do this string/str conversion dance and we're always going to use the metric property. Might make some of these logic blocks less busy but would only work for metrics in the agent.